netty/src/main/java/io/grpc/netty/NettyClientHandler.java

```
@@ -591,6 +613,29 @@ private void createStream(CreateStreamCommand command, ChannelPromise promise)
       return;
     }
 
+    String authority = getAuthorityPseudoHeader(command.headers());
+    if (authority != null) {
+      Status authorityVerificationStatus = peerVerificationResults.get(authority);
+      if (authorityVerificationStatus == null && attributes != null
+          && attributes.get(GrpcAttributes.ATTR_AUTHORITY_VERIFIER) != null) {
+        authorityVerificationStatus = attributes.get(GrpcAttributes.ATTR_AUTHORITY_VERIFIER)
+            .verifyAuthority(((GrpcHttp2OutboundHeaders) command.headers()).getAuthority());
+        peerVerificationResults.put(authority, authorityVerificationStatus);
+      }
+      if (authorityVerificationStatus != null && !authorityVerificationStatus.isOk()) {
```

This is for security, so we need to fail closed. If there is no verifier, then it isn't allowed.

---

What should be the behavior for Plain text negotiation handler such as used in the NettyClientTransport [huffman encoding test](https://github.com/grpc/grpc-java/blob/5a7f350537bbdbba3e3f49de53d298d99b5c1913/netty/src/test/java/io/grpc/netty/NettyClientTransportTest.java#L611)?

---

> As a subsequent work I assume.

No, we'd need it at the same time, as this would be breaking existing code potentially, and there isn't any security concern. Even if we do it as a separate PR before the next release, we could still potentially break Google.

We need to look through all existing client protocol negotiators and consider updating them. But that's not as bad as it sounds:

* ProtocolNegotiators.ProxyProtocolNegotiationHandler. Delegates to another negotator. No need to change
* ProtocolNegotiators.ClientTlsHandler. You already are changing
* ProtocolNegotiators.WaitUntilActiveHandler. This [is plaintext](https://github.com/grpc/grpc-java/blob/60f6ea7b8eec1692fc877ec82dd67d6bb888f706/netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java#L980) (and [plaintext upgrade](https://github.com/grpc/grpc-java/blob/60f6ea7b8eec1692fc877ec82dd67d6bb888f706/netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java#L783))... but also used outside of plaintext. And we don't want to rely on ["Later handlers are expected to overwrite this."](https://github.com/grpc/grpc-java/blob/60f6ea7b8eec1692fc877ec82dd67d6bb888f706/netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java#L1031).
* SecurityProtocolNegotiators.ClientSecurityProtocolNegotiator. Delegates to ClientTlsHandler, so it will be secure by default. But this is actually the xds implementation that we'll give its own behavior. Can be done later
* TsiHandshakeHandler. Part of ALTS. Don't update
* S2AProtocolNegotiatorFactory.S2AProtocolNegotiationHandler. Don't update

So literally only plaintext needs extra work right now.

> After the changes to disallow authority override without having a verifier in the attributes, the FakeControlPlaneIntegrationTest which doesn't use Tls but plain text transport is failing.

Exactly. We need to fix plaintext.

---

This should be enough for plaintext.

```diff
diff --git a/netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java b/netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java
index 33a8e6e1a..f8ed7a011 100644
--- a/netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java
+++ b/netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java
@@ -836,7 +836,9 @@ final class ProtocolNegotiators {
     public ChannelHandler newHandler(GrpcHttp2ConnectionHandler grpcHandler) {
       ChannelHandler upgradeHandler =
           new Http2UpgradeAndGrpcHandler(grpcHandler.getAuthority(), grpcHandler);
-      return new WaitUntilActiveHandler(upgradeHandler, grpcHandler.getNegotiationLogger());
+      ChannelHandler plaintextHandler =
+          new PlaintextHandler(upgradeHandler, grpcHandler.getNegotiationLogger());
+      return new WaitUntilActiveHandler(plaintextHandler, grpcHandler.getNegotiationLogger());
     }
 
     @Override
@@ -1033,7 +1035,9 @@ final class ProtocolNegotiators {
     @Override
     public ChannelHandler newHandler(GrpcHttp2ConnectionHandler grpcHandler) {
       ChannelHandler grpcNegotiationHandler = new GrpcNegotiationHandler(grpcHandler);
-      ChannelHandler activeHandler = new WaitUntilActiveHandler(grpcNegotiationHandler,
+      ChannelHandler plaintextHandler =
+          new PlaintextHandler(grpcNegotiationHandler, grpcHandler.getNegotiationLogger());
+      ChannelHandler activeHandler = new WaitUntilActiveHandler(plaintextHandler,
           grpcHandler.getNegotiationLogger());
       return activeHandler;
     }
@@ -1047,6 +1051,22 @@ final class ProtocolNegotiators {
     }
   }
 
+  static final class PlaintextHandler extends ProtocolNegotiationHandler {
+    PlaintextHandler(ChannelHandler next, ChannelLogger negotiationLogger) {
+      super(next, negotiationLogger);
+    }
+
+    @Override
+    protected void protocolNegotiationEventTriggered(ChannelHandlerContext ctx) {
+      ProtocolNegotiationEvent existingPne = getProtocolNegotiationEvent();
+      Attributes attrs = existingPne.getAttributes().toBuilder()
+          .set(GrpcAttributes.ATTR_AUTHORITY_VERIFIER, (authority) -> Status.OK)
+          .build();
+      replaceProtocolNegotiationEvent(existingPne.withAttributes(attrs));
+      fireProtocolNegotiationEvent(ctx);
+    }
+  }
+
   /**
    * Waits for the channel to be active, and then installs the next Handler.  Using this allows
    * subsequent handlers to assume the channel is active and ready to send.  Additionally, this a
```

---

Thanks.

---

> * SecurityProtocolNegotiators.ClientSecurityProtocolNegotiator. Delegates to ClientTlsHandler, so it will be secure by default. But this is actually the xds implementation that we'll give its own behavior. Can be done later

Actually it delegates to ClientSecurityHandler not to ClientTlsHandler, so overriding authority via call option will fail for this handler. I hope it is still ok.

---

ClientSecurityHandler delegates to ClientTlsHandler, but I'm not surprised it wasn't obvious to you:
https://github.com/grpc/grpc-java/blob/892144dcac18a1cedbdc4cddecefb41193a965fc/xds/src/main/java/io/grpc/xds/internal/security/SecurityProtocolNegotiators.java#L225

But even if it didn't, yeah, it would be okay. We don't really expect existing xds users to override authority. It isn't something we are well-supporting in xds routing/config selector today.